include/zephyr/dsp/types.h

```diff
@@ -0,0 +1,71 @@
+/* Copyright (c) 2022 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef INCLUDE_ZEPHYR_DSP_TYPES_H_
+#define INCLUDE_ZEPHYR_DSP_TYPES_H_
+
+#include <stdint.h>
+
+/**
+ * @addtogroup math_dsp
+ * @{
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @typedef q7_t
+ * @brief 8-bit fractional data type in 1.7 format.
+ */
+typedef int8_t q7_t;
+
+/**
+ * @typedef q15_t
+ * @brief 16-bit fractional data type in 1.15 format.
+ */
+typedef int16_t q15_t;
+
+/**
+ * @typedef q31_t
+ * @brief 32-bit fractional data type in 1.31 format.
+ */
+typedef int32_t q31_t;
+
+/**
+ * @typedef q63_t
+ * @brief 64-bit fractional data type in 1.63 format.
+ */
+typedef int64_t q63_t;
+
+/**
+ * @typedef float16_t
+ * @brief 16-bit floating point type definition.
+ */
+#if defined(CONFIG_FP16)
+typedef __fp16 float16_t;
```

Comment:
Just a note to myself (or anyone who might be interested in maintaining this in the future) that `CONFIG_FP16` is currently an ARM-specific config and `__fp16` is an ARM-specific type.

The `CONFIG_FP16` Kconfig should eventually be moved to the same level as `CONFIG_FPU` and the architecture-agnostic ISO `_Float16` type support should be implemented.

Reply:
Without knowing the details myself: is there a reason for not using the ISO _Float16 already now then? Is it semantically different from __fp16?

Reply:
On ARM, the `__fp16` type can represent a half-precision floating point value in either IEEE 754-2008 format or the "ARM alternative format" depending on the specified compiler options -- we do not want to take away that capability for now.

Reply:
https://gcc.gnu.org/onlinedocs/gcc/Half-Precision.html